File: internal/scrapers/wanted.go
// Package scrapers는 bronze 수집기들이다. 각 수집기는 Scraper 시그니처를 만족하며,
// 파이프라인에 주입된다. 실패는 오류로 반환하고 격리 판단은 파이프라인이 한다.
package scrapers

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"math/rand"
	"net/http"
	"net/url"
	"time"

	"github.com/Chae-JS/mentoai/internal/config"
	"github.com/Chae-JS/mentoai/internal/domain"
)

// wanted는 Wanted 채용 JSON API를 사용한다.
type Wanted struct {
	baseURL      string
	groupID      string
	jobIDs       string
	maxItems     int
	delaySeconds float64
	http         *http.Client
	rand         *rand.Rand
}

// NewWanted는 설정에서 Wanted 수집기를 만든다.
func NewWanted(s config.Settings) *Wanted {
	return &Wanted{
		baseURL:      s.WantedBaseURL,
		groupID:      s.WantedJobGroupID,
		jobIDs:       s.WantedJobIDs,
		maxItems:     s.ScrapeMaxItems,
		delaySeconds: s.ScrapeDelaySeconds,
		http:         &http.Client{Timeout: 15 * time.Second},
		rand:         rand.New(rand.NewSource(time.Now().UnixNano())),
	}
}

const userAgent = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) " +
	"AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"

type wantedListResponse struct {
	Data []struct {
		ID *int64 `json:"id"`
	} `json:"data"`
	Links struct {
		Next *string `json:"next"`
	} `json:"links"`
}

type wantedDetailResponse struct {
	Data struct {
		Job json.RawMessage `json:"job"`
	} `json:"data"`
}

// fetchJSON은 JSON API를 호출하고 응답 본문을 디코딩한다.
func fetchJSON(ctx context.Context, client *http.Client, endpoint string, out any) (int, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, endpoint, nil)
	if err != nil {
		return 0, err
	}
	req.Header.Set("User-Agent", userAgent)

	resp, err := client.Do(req)
	if err != nil {
		return 0, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		io.Copy(io.Discard, resp.Body)
		return resp.StatusCode, nil
	}
	return resp.StatusCode, json.NewDecoder(resp.Body).Decode(out)
}

// Scrape는 Wanted 공고를 수집해 bronze 레코드로 반환한다.
func (w *Wanted) Scrape(ctx context.Context) ([]domain.RawRecord, error) {
	records := []domain.RawRecord{}
	ids, err := w.fetchJobIDs(ctx)
	if err != nil {
		return nil, err
	}

	logf("wanted: 수집 대상 %d건", len(ids))
	for _, jobID := range ids {
		select {
		case <-ctx.Done():
			return records, ctx.Err()
		default:
		}

		raw, err := w.fetchDetail(ctx, jobID)
		if err != nil {
			return nil, err
		}
		if raw != nil {
			records = append(records, domain.RawRecord{
				Source:      "wanted",
				SourceID:    fmt.Sprint(jobID),
				CollectedAt: nowRFC3339(),
				Payload:     raw,
			})
		}
		sleepJitter(ctx, w.rand, w.delaySeconds)
	}
	return records, nil
}

func (w *Wanted) fetchJobIDs(ctx context.Context) ([]int64, error) {
	var ids []int64
	offset := 0
	for len(ids) < w.maxItems {
		q := url.Values{}
		q.Set("job_group_id", w.groupID)
		q.Set("job_ids", w.jobIDs)
		q.Set("country", "kr")
		q.Set("job_sort", "job.popularity_order")
		q.Set("years", "-1")
		q.Set("locations", "all")
		q.Set("limit", "20")
		q.Set("offset", fmt.Sprint(offset))

		var data wantedListResponse
		status, err := fetchJSON(ctx, w.http, w.baseURL+"/api/chaos/navigation/v1/results?"+q.Encode(), &data)
		if err != nil {
			return nil, fmt.Errorf("wanted 리스트 요청 실패: %w", err)
		}
		if status != http.StatusOK {
			logf("wanted 리스트 요청 실패: %d", status)
			break
		}
		if len(data.Data) == 0 {
			break
		}
		for _, job := range data.Data {
			if job.ID != nil {
				ids = append(ids, *job.ID)
			}
		}
		if data.Links.Next == nil {
			break
		}
		offset += 20
		sleepJitter(ctx, w.rand, 0.4)
	}
	if len(ids) > w.maxItems {
		ids = ids[:w.maxItems]
	}
	return ids, nil
}

func (w *Wanted) fetchDetail(ctx context.Context, jobID int64) ([]byte, error) {
	endpoint := fmt.Sprintf("%s/api/chaos/jobs/v4/%d/details", w.baseURL, jobID)
	var data wantedDetailResponse
	status, err := fetchJSON(ctx, w.http, endpoint, &data)
	if err != nil {
		return nil, fmt.Errorf("wanted 상세 요청 실패 (ID: %d): %w", jobID, err)
	}
	if status == http.StatusNotFound {
		logf("wanted 공고 삭제/비공개 (ID: %d)", jobID)
		return nil, nil
	}
	if status != http.StatusOK {
		logf("wanted 상세 요청 실패: %d (ID: %d)", status, jobID)
		return nil, nil
	}
	if len(data.Data.Job) == 0 {
		return nil, nil
	}
	return data.Data.Job, nil
}

File: src/mentoai/scrapers/wanted.py
import asyncio
import logging
import random
from datetime import UTC, datetime
from typing import Any

import httpx

from mentoai.config import get_settings

logger = logging.getLogger(__name__)

LIST_API = "/api/chaos/navigation/v1/results"
DETAIL_API = "/api/chaos/jobs/v4/{job_id}/details"

HEADERS = {
    "User-Agent": (
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) "
        "AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
    ),
}


async def fetch_job_ids(
    client: httpx.AsyncClient, group_id: str, job_ids: str, max_items: int
) -> list[int]:
    ids: list[int] = []
    offset = 0
    while len(ids) < max_items:
        response = await client.get(
            LIST_API,
            params={
                "job_group_id": group_id,
                "job_ids": job_ids,
                "country": "kr",
                "job_sort": "job.popularity_order",
                "years": "-1",
                "locations": "all",
                "limit": "20",
                "offset": offset,
            },
        )
        if response.status_code != 200:
            logger.warning("wanted 리스트 요청 실패: %s", response.status_code)
            break
        data = response.json()
        jobs = data.get("data", [])
        if not jobs:
            break
        ids.extend(job["id"] for job in jobs if job.get("id") is not None)
        if not data.get("links", {}).get("next"):
            break
        offset += 20
        await asyncio.sleep(random.uniform(0.4, 0.8))
    return ids[:max_items]


async def fetch_detail(client: httpx.AsyncClient, job_id: int) -> dict[str, Any] | None:
    response = await client.get(DETAIL_API.format(job_id=job_id))
    if response.status_code == 404:
        logger.info("wanted 공고 삭제/비공개 (ID: %s)", job_id)
        return None
    if response.status_code != 200:
        logger.warning("wanted 상세 요청 실패: %s (ID: %s)", response.status_code, job_id)
        return None
    return response.json().get("data", {}).get("job") or None


async def scrape() -> list[dict[str, Any]]:
    """Wanted 공고를 수집해 bronze 레코드로 반환한다."""
    settings = get_settings()
    records: list[dict[str, Any]] = []
    async with httpx.AsyncClient(
        base_url=settings.wanted_base_url,
        headers={**HEADERS, "Referer": settings.wanted_base_url},
        timeout=15,
        follow_redirects=True,
    ) as client:
        ids = await fetch_job_ids(
            client,
            settings.wanted_job_group_id,
            settings.wanted_job_ids,
            settings.scrape_max_items,
        )
        logger.info("wanted: 수집 대상 %d건", len(ids))
        for job_id in ids:
            raw = await fetch_detail(client, job_id)
            if raw:
                records.append(
                    {
                        "source": "wanted",
                        "source_id": str(job_id),
                        "collected_at": datetime.now(UTC),
                        "payload": raw,
                    }
                )
            await asyncio.sleep(settings.scrape_delay_seconds + random.uniform(0, 0.3))
    return records
